File: dynamic dispatch.cpp
//
// Created by nived on 07-01-2026.
//
#include <iostream>
using namespace std;

class Payment {
public:
    virtual void pay(int amount) {
        cout << "Base payment\n";
    }
    virtual ~Payment() {}
};

class CardPayment : public Payment {
public:
    void pay(int amount) override {
        cout << "Card payment\n";
    }
};

int main() {
    Payment* p = new CardPayment();
    p->pay(100);   // Card payment
}

File: dynamic_dispatch.java
abstract class Payment {
    abstract void pay(int amount);
}

class CardPayment extends Payment {
    void pay(int amount) {
        System.out.println("Card payment");
    }
}

public class dynamic_dispatch {
    public static void main(String[] args) {
        Payment p = new CardPayment();
        p.pay(100); // runtime dispatch
    }
}
